compiler/rustc_trait_selection/src/solve/eval_ctxt/select.rs

```diff
@@ -135,44 +131,30 @@ impl<'tcx> InferCtxtSelectExt<'tcx> for InferCtxt<'tcx> {
 impl<'tcx> EvalCtxt<'_, 'tcx> {
     fn compute_canonical_trait_candidates(
         &mut self,
-        canonical_input: Canonical<'tcx, QueryInput<'tcx, ty::TraitPredicate<'tcx>>>,
+        canonical_input: CanonicalInput<'tcx>,
     ) -> Vec<Candidate<'tcx>> {
-        let intercrate = match self.search_graph.solver_mode() {
-            SolverMode::Normal => false,
-            SolverMode::Coherence => true,
-        };
-        let (canonical_infcx, input, var_values) = self
-            .tcx()
-            .infer_ctxt()
-            .intercrate(intercrate)
-            .with_next_trait_solver(true)
-            .with_opaque_type_inference(canonical_input.value.anchor)
-            .build_with_canonical(DUMMY_SP, &canonical_input);
-
-        let mut ecx = EvalCtxt {
-            infcx: &canonical_infcx,
-            var_values,
-            predefined_opaques_in_body: input.predefined_opaques_in_body,
-            max_input_universe: canonical_input.max_universe,
-            search_graph: &mut self.search_graph,
-            nested_goals: NestedGoals::new(),
-            tainted: Ok(()),
-            inspect: ProofTreeBuilder::new_noop(),
-        };
-
-        for &(key, ty) in &input.predefined_opaques_in_body.opaque_types {
-            ecx.insert_hidden_type(key, input.goal.param_env, ty)
-                .expect("failed to prepopulate opaque types");
-        }
-
-        let candidates = ecx.assemble_and_evaluate_candidates(input.goal);
-
-        // We don't need the canonicalized context anymore
-        if input.anchor != DefiningAnchor::Error {
-            let _ = canonical_infcx.take_opaque_types();
-        }
-
-        candidates
+        // FIXME: This does not actually record the canonical goal in the search graph,
+        // so this may report false positives for cycles. This seems problematic, and
+        // should be fixed.
+        EvalCtxt::enter_canonical(
+            self.tcx(),
+            self.search_graph(),
+            canonical_input,
+            // FIXME: This is wrong, idk if we even want to track stuff here.
```

Comment:
for debugging it would be good. I don't know how to best impl that though because select is kind of weird from that POV. ideally this `enter_canonical` doesn't but all the nested goals do :thinking: can keep this open for later, don't need the proof graph of this for diagnostics or auto traits, so it only negatively impacts debugging